### PR TITLE
[WIP]NULS 2.0 integration

### PR DESCRIPTION
**Pull request by ccccbjcn**
## Description

add nuls 2.0 into TrustWallet.

## Testing instructions

1.add NULS 2.0 address.
2.add NULS 2.0 account model transaction and signature.
3.add address and transaction unit test.

## Types of changes

New feature (non-breaking change which adds functionality) 

## Checklist

- [x] Prefix PR title with `[WIP]` if necessary.
- [x] Add tests to cover changes as needed.
- [x] Update documentation as needed.

### CODE REVIEW

File: coins.json [hunk 1]
```diff
@@ -705,6 +705,23 @@
...
+        "symbol": "NULS",
+        "decimals": 8,
+        "blockchain": "NULS",
+        "derivationPath": "m/44'/8964'/0'/0/0",
+        "curve": "secp256k1",
+        "publicKeyType": "secp256k1",
+        "explorer": "https://nuls.world/transactions/",
+        "info": {
+            "url": "https://nuls.io",
+            "client": "https://github.com/trezor/blockbook",
```
Comment (hewigovens) [inline coins.json]: should not be blockbook

File: coins.json [hunk 2]
```diff
@@ -705,6 +705,23 @@
...
+            "clientPublic": "",
```
Comment (hewigovens) [inline coins.json]: please also provider this: https://developer.trustwallet.com/wallet-core/rpc-requirements

File: include/TrustWalletCore/TWNULSAddress.h [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+//
+// This file is part of Trust. The full Trust copyright notice, including
+// terms governing use, modification, and redistribution, is contained in the
+// file LICENSE at the root of the source code distribution tree.
+
+#pragma once
+
+#include "TWBase.h"
+#include "TWData.h"
+#include "TWHRP.h"
```
Comment (hewigovens) [inline include/TrustWalletCore/TWNULSAddress.h]: no hrp in NULS

File: include/TrustWalletCore/TWNULSAddress.h [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+TW_EXPORT_STATIC_METHOD
+bool TWNULSAddressIsValidString(TWString *_Nonnull string);
+
+/// Creates an address from a string representaion.
+TW_EXPORT_STATIC_METHOD
+struct TWNULSAddress *_Nullable TWNULSAddressCreateWithString(TWString *_Nonnull string);
+
+/// Creates an address from a raw data representaion.
+TW_EXPORT_STATIC_METHOD
+struct TWNULSAddress *_Nullable TWNULSAddressCreateWithData(TWData *_Nonnull data);
```
Comment (hewigovens) [inline include/TrustWalletCore/TWNULSAddress.h]: remove this method if you don't use it

File: include/TrustWalletCore/TWNULSSigner.h [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+
+TW_EXTERN_C_BEGIN
+
+/// Helper class to sign NULS transactions.
+TW_EXPORT_CLASS
+struct TWNULSSigner;
+
+/// Make a transaction plan
+//TW_EXPORT_STATIC_METHOD
+//TW_NULS_Proto_TransactionPlan TWNULSSignerPlan(TW_NULS_Proto_TransactionPurpose purpose);
```
Comment (hewigovens) [inline include/TrustWalletCore/TWNULSSigner.h]: remove not used code

File: js/tests/utils/CoinAddressDerivation.test.ts [hunk 6]
```diff
@@ -0,0 +1,77 @@
+import { expect } from 'chai';
```
Comment (hewigovens) [inline js/tests/utils/CoinAddressDerivation.test.ts]: we don't support js now

File: src/Coin.cpp [hunk 7]
```diff
@@ -40,6 +40,11 @@
 #include "Wanchain/Address.h"
 #include "Waves/Address.h"
 #include "Zcash/TAddress.h"
+#include "NULS/Address.h"
+#include "Bravo/Address.h"
```
Comment (hewigovens) [inline src/Coin.cpp]: a few duplicated #include

File: src/NULS/Address.cpp [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+#include "../BinaryCoding.h"
+#include "../HexCoding.h"
+
+using namespace TW::NULS;
+
+bool Address::isValid(const std::string& string) {
+    if (string.empty()) {
+        return false;
+    }
+    std::string prefix = "NULSd";
```
Comment (hewigovens) [inline src/NULS/Address.cpp]: use constant variable

File: src/NULS/Address.cpp [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+    for (int i = 0; i < 23; ++i) {
+        checkSum ^= decoded[i];
+    }
+
+    return decoded[23] == checkSum;
+}
+
+Address::Address(const TW::PublicKey& publicKey) {
+    // Main-Net chainID
+    bytes[0] = 0x01;
```
Comment (hewigovens) [inline src/NULS/Address.cpp]: constant variables

File: src/NULS/Address.cpp [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+    for (int i = 0; i < 23; ++i) {
+        checkSum ^= bytes[i];
+    }
+    bytes[23] = checkSum;
+}
+
+Address::Address(const std::string& string) {
+    std::string prefix = "NULSd";
+    std::string address = string.substr(prefix.length(), string.length() - prefix.length());   
+    const auto decoded = Base58::bitcoin.decode(address);
```
Comment (hewigovens) [inline src/NULS/Address.cpp]: call Address::isValid to verify the checksum

File: src/NULS/Address.cpp [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+
+    uint8_t checkSum = 0x00;
+    for (int i = 0; i < 23; ++i) {
+        checkSum ^= bytes[i];
+    }
+
+    return bytes[23] == checkSum;
+}
+
+TW::PrivateKey Address::importHexPrivateKey(std::string hexPrivateKey) {
```
Comment (hewigovens) [inline src/NULS/Address.cpp]: remove this, we only support import via recovery phrases

File: src/NULS/Address.cpp [hunk 12]
```diff
@@ -0,0 +1,123 @@
...
+        throw std::invalid_argument("Invalid private key");
+    }
+    }
+
+    auto key = PrivateKey(data);
+    return key;
+}
+
+std::string Address::string() const {
+    std::string prefix = "NULSd";
```
Comment (hewigovens) [inline src/NULS/Address.cpp]: can you use constant variable?

File: src/NULS/BinaryCoding.h [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+
+static inline void serializerRemark(std::string& remark, std::vector<uint8_t>& data) {
+    encodeVarInt(remark.length(), data);
+    std::copy(remark.begin(), remark.end(), std::back_inserter(data));
+}
+
+static inline void serializerInput(std::vector<Proto::TransactionCoinFrom>& inputs,
+    std::vector<uint8_t>& data) {
+    encodeVarInt(1, data);  //there is one coinFrom
+    for (const auto& input : inputs) {
```
Comment (hewigovens) [inline src/NULS/BinaryCoding.h]: if only one, remove for loop

File: src/NULS/BinaryCoding.h [hunk 14]
```diff
@@ -0,0 +1,99 @@
...
+        append(data, nonce);
+        data.push_back(static_cast<uint8_t>(input.locked()));
+        break;
+    }
+}
+
+static inline void serializerOutput(std::vector<Proto::TransactionCoinTo>& outputs,
+    std::vector<uint8_t>& data) {
+    encodeVarInt(1, data); //there is one coinTo
+    for (const auto& output : outputs) {
```
Comment (hewigovens) [inline src/NULS/BinaryCoding.h]: remove for loop

File: src/NULS/BinaryCoding.h [hunk 15]
```diff
@@ -0,0 +1,99 @@
...
+    std::copy(pubKey.bytes.begin(), pubKey.bytes.end(), std::back_inserter(transactionSignature));
+    auto signature = privateKey.signAsDER(txHash, TWCurve::TWCurveSECP256k1);
+    encodeVarInt(signature.size(), transactionSignature);
+    std::copy(signature.begin(), signature.end(), std::back_inserter(transactionSignature));
+    return transactionSignature;
+}
+
+/// Loads a `uint256_t` from Protobuf bytes (which are wrongly represented as
+/// std::string).(lsb first)
+inline uint256_t load_lsb(const std::string& data) {
```
Comment (hewigovens) [inline src/NULS/BinaryCoding.h]: what's lsb? can we reuse methods in `uint256.h`? like do a reverse

File: src/NULS/TransactionBuilder.h [hunk 16]
```diff
@@ -0,0 +1,51 @@
...
+
+#pragma once
+
+#include "../proto/NULS.pb.h"
+
+#include <algorithm>
+
+namespace TW::NULS {
+
+struct TransactionBuilder {
```
Comment (hewigovens) [inline src/NULS/TransactionBuilder.h]: do we still need this?

File: src/interface/TWNULSAddress.cpp [hunk 17]
```diff
@@ -0,0 +1,58 @@
...
+/// Determines if the string is a valid NULS address.
+bool TWNULSAddressIsValidString(TWString* _Nonnull string) {
+    auto s = reinterpret_cast<const std::string*>(string);
+    return Address::isValid(*s);
+}
+
+/// Creates an address from a string representaion.
+struct TWNULSAddress* _Nullable TWNULSAddressCreateWithString(TWString* _Nonnull string) {
+    auto s = reinterpret_cast<const std::string*>(string);
+    const auto address = Address(*s);
```
Comment (hewigovens) [inline src/interface/TWNULSAddress.cpp]: try catch here

File: src/interface/TWNULSAddress.cpp [hunk 18]
```diff
@@ -0,0 +1,58 @@
...
+    return new TWNULSAddress{std::move(address)};
+}
+
+struct TWNULSAddress* _Nullable TWNULSAddressCreateWithData(TWData* _Nonnull data) {
```
Comment (hewigovens) [inline src/interface/TWNULSAddress.cpp]: remove createWithData if you don't need it

File: src/proto/NULS.proto [hunk 19]
```diff
@@ -0,0 +1,66 @@
+syntax = "proto3";
+
+package TW.NULS.Proto;
+option java_package = "wallet.core.jni.proto";
+
+message TransactionCoinFrom {
```
Comment (hewigovens) [inline src/proto/NULS.proto]: maybe CoinFrom -> Input?

File: src/proto/NULS.proto [hunk 20]
```diff
@@ -0,0 +1,66 @@
...
+    uint32 idassets_id = 3;
+    //tranaction out amount (128-bit number)
+    bytes idamount = 4;
+    //8 bytes
+    bytes nonce = 5;
+    //lock status: 1 locked; 0 unlocked
+    uint32 locked = 6;
+}
+
+message TransactionCoinTo {
```
Comment (hewigovens) [inline src/proto/NULS.proto]: CoinTo -> Output

File: src/proto/NULS.proto [hunk 21]
```diff
@@ -0,0 +1,66 @@
...
+
+package TW.NULS.Proto;
+option java_package = "wallet.core.jni.proto";
+
+message TransactionCoinFrom {
+    string from_address = 1;
+    uint32 assets_chainid = 2;
+    uint32 idassets_id = 3;
+    //tranaction out amount (128-bit number)
+    bytes idamount = 4;
```
Comment (hewigovens) [inline src/proto/NULS.proto]:
```suggestion
    bytes id_amount = 4;
```

File: src/proto/NULS.proto [hunk 22]
```diff
@@ -0,0 +1,66 @@
...
+    //lock status: 1 locked; 0 unlocked
+    uint32 locked = 6;
+}
+
+message TransactionCoinTo {
+    string to_address = 1;
+    uint32 assets_chainid = 2;
+    uint32 idassets_id = 3;
+    // tranaction amount (128-bit number)
+    bytes idamount = 4;
```
Comment (hewigovens) [inline src/proto/NULS.proto]:
```suggestion
    bytes id_amount = 4;
```

File: src/proto/NULS.proto [hunk 23]
```diff
@@ -0,0 +1,66 @@
...
+    //8 bytes
+    bytes nonce = 5;
+    //lock status: 1 locked; 0 unlocked
+    uint32 locked = 6;
+}
+
+message TransactionCoinTo {
+    string to_address = 1;
+    uint32 assets_chainid = 2;
+    uint32 idassets_id = 3;
```
Comment (hewigovens) [inline src/proto/NULS.proto]:
```suggestion
    uint32 assets_id = 3;
```

File: src/proto/NULS.proto [hunk 24]
```diff
@@ -0,0 +1,66 @@
...
+    repeated TransactionCoinTo outputs = 6;
+
+    repeated Signature tx_sigs = 7;
+    uint32 hash = 8;
+}
+
+// Input data necessary to create a signed order.
+message SigningInput {
+    bytes private_key = 1;
+    string from_address = 2;
```
Comment (hewigovens) [inline src/proto/NULS.proto]: you can get from address from private key, so you don't have to pass it

File: tests/NULS/AddressTests.cpp [hunk 25]
```diff
@@ -0,0 +1,77 @@
...
+
+#include "HexCoding.h"
+#include "PrivateKey.h"
+#include <gtest/gtest.h>
+
+using namespace TW;
+using namespace TW::NULS;
+
+
+TEST(NULSAddress, StaticInvalid) {
```
Comment (hewigovens) [inline tests/NULS/AddressTests.cpp]: can you combine a single address validation tests?

File: tests/NULS/AddressTests.cpp [hunk 26]
```diff
@@ -0,0 +1,77 @@
...
+    const auto address = Address("NULSd6HgbwcM8wz48f6UkFYHLVriT1L81X9z2");
+    ASSERT_TRUE(address.type() == 1);
+}
+
+TEST(NULSAddress, FromString) {
+    const auto address = Address("NULSd6HgbwcM8wz48f6UkFYHLVriT1L81X9z2");
+    ASSERT_EQ(address.string(), "NULSd6HgbwcM8wz48f6UkFYHLVriT1L81X9z2");
+}
+
+TEST(NULSAddress, FromData) {
```
Comment (hewigovens) [inline tests/NULS/AddressTests.cpp]: remove it

File: tests/NULS/SignerTests.cpp [hunk 27]
```diff
@@ -0,0 +1,64 @@
...
+    input.set_timestamp(0x5d8885f8);
+    return input;
+}
+
+TEST(NULSSigner, Sign) {
+    auto input = makeInput("9ce21dad67e0f0af2599b41b515a7f7018059418bab892a7b68f283d489abc4b",
+                           "NULSd6Hgj7ZoVgsPN9ybB4C1N2TbvkgLc8Z9H", "NULSd6Hgied7ym6qMEfVzZanMaa9qeqA6TZSe",
+                           10000000, 1, 1, "0000000000000000", "", 100000000);
+    auto signer = NULS::Signer(input);
+    std::vector<uint8_t> signature = signer.sign();
```
Comment (hewigovens) [inline tests/NULS/SignerTests.cpp]:
```suggestion
    Data signature = signer.sign();
```
you can use `Data` typealias

File: tests/NULS/deriveAddressTests.cpp [hunk 28]
```diff
@@ -0,0 +1,76 @@
...
+#include "DerivationPath.h"
+
+#include <gtest/gtest.h>
+
+using namespace TW;
+using namespace TW::NULS;
+
+// https://github.com/kiss1987f4/blockbook/blob/master/bchain/coins/nuls/nulsparser_test.go#L347
+
+TEST(NULSAddress, DeriveKeyFromXPrv) {
```
Comment (hewigovens) [inline tests/NULS/deriveAddressTests.cpp]: remove this file

**Comment by hewigovens** [comment]
please also add swift/kotlin tests, fix CI and codacy warnings

File: src/proto/NULS.proto (hunk 19 above)
Comment (ccccbjcn) [reply src/proto/NULS.proto]: CoinFrom is a class type defined in NULS 2.0. If it is confuse, iI will change it to Input.

File: src/proto/NULS.proto (hunk 21 above)
Comment (ccccbjcn) [reply src/proto/NULS.proto]: what is this item you want to comment?

File: src/proto/NULS.proto (hunk 24 above)
Comment (ccccbjcn) [reply src/proto/NULS.proto]: What is the api to calll to get private key from address?
I think the code would run in wallet core of  TW APP.

Comment (hewigovens) [reply src/proto/NULS.proto]: you can derive address from your private key, that should be your `from_address`?

Comment (ccccbjcn) [reply src/proto/NULS.proto]: I want to get private key to signature transaction.
How could  I get private key from from_address which is carried in protobuf message?

**Comment by ccccbjcn** [comment]
> please also add swift/kotlin tests, fix CI and codacy warnings

Per the legacy nuls 1.0 code, i did not get the swift/kotlin code and test.  How the legacy code was merged? I will feedback to nuls 2.0 core team.

**Comment by hewigovens** [comment]
We have more formal process to add integrations, beside, you have to add a mainnet tx test

**Comment by ccccbjcn** [comment]
> We have more formal process to add integrations, beside, you have to add a mainnet tx test

What the test cases need to be added?

File: src/proto/NULS.proto (hunk 24 above)
Comment (ccccbjcn) [reply src/proto/NULS.proto]: @hewigovens  please reply how to getting private key. I read token proto files defining private key attribute.

Comment (hewigovens) [reply src/proto/NULS.proto]: do you mean this?
```c++
// Proto::SigningInput &input
PrivateKey(Data(input.private_key().begin(), input.private_key().end()));
```

File: include/TrustWalletCore/TWNULSAddress.h (hunk 4 above)
Comment (ccccbjcn) [reply include/TrustWalletCore/TWNULSAddress.h]: resolve it.

File: src/NULS/Address.cpp (hunk 11 above)
Comment (ccccbjcn) [reply src/NULS/Address.cpp]: reserve it.

File: src/proto/NULS.proto (hunk 24 above)
Comment (ccccbjcn) [reply src/proto/NULS.proto]: Yes, how to get private key from proto message or other ways.

**Pull request closed without merging**